### PR TITLE
DCP Architecture Team charter

### PR DESCRIPTION
**Pull request by mweiden**
**Approved at the 10/11/18 DCP PM meeting**

This PR proposes the Architecture++ charter.

Please see charter for more information.

### CODE REVIEW

File: charters/Architecture/charter.md
Comment (brianraymor) [inline charters/Architecture/charter.md]: Perhaps `s/Architecture++/DCP Architecture/` - maybe even _DCP Architecture Team_, the playful **++** notation in the presentation was intended to convey the next small increment of an existing group. Of course, if the technical leads enjoy **++**, then by all means keep it.

File: charters/Architecture/charter.md
Comment (brianraymor) [inline charters/Architecture/charter.md]: Is it clear to other reviewers that Architecture is the **final** approver for technical charters?

File: charters/Architecture/charter.md
Comment (brianraymor) [inline charters/Architecture/charter.md]: Is it clear to other reviewers that Architecture is the **final** approver for cross-component RFC(s)?

File: charters/Architecture/charter.md
Comment (brianraymor) [inline charters/Architecture/charter.md]: There has been a slight preference to refer to DCP _projects_ rather than _components_.

File: charters/Architecture/charter.md
Comment (brianraymor) [inline charters/Architecture/charter.md]: How about:
_Coordinate with DCP PM on emerging tactical requests needing to be addressed in Sprint planning priorities_

File: charters/Architecture/charter.md
Comment (sampierson) [reply charters/Architecture/charter.md]: Slight preference from whom?  We have historically used the term "component" to refer to DCP subassemblies.  The term "Project" refers to a initiative or scope of work, a more vague term, which could mean "build a component" or "hold a meeting", whereas "component" implies a concrete deliverable.  The Architecture Team was created to address architectural issues that cut across components.  You may wish for it to also oversee other types of DCP projects, but this would be in addition to, not instead of, components.

Comment (brianraymor) [reply charters/Architecture/charter.md]: See the [extended conversation in the charter process PR](https://github.com/HumanCellAtlas/dcp-community/pull/1#discussion_r204456321) about products, projects, components, modules, etc.

File: charters/Architecture/charter.md
Comment (brianraymor) [inline charters/Architecture/charter.md]: Is **annual** a good cadence for the DCP Architecture RFC from the perspective of the technical leads? I'm curious because DCP PM still needs to explore/publish refinements to the planning process for 2019. 

And regarding **roadmap**, we still need to agree on a common term for the planning artifact that _identifies and expresses customer needs as a set of themes or goals_ similar in nature to [Rust](https://blog.rust-lang.org/2018/03/12/roadmap.html) and other open source projects that rally their communities. 

How about:

* Publish an Architecture RFC documenting technical debt that must be addressed in DCP PM product plans

Comment (tburdett) [reply charters/Architecture/charter.md]: Regarding technical debt: I've been trying to identify the reason for some of my discomfort here, and I think it is down to the idea of universal DCP-wide technical debt. Individual charters will have differing requirements and therefore will be more or less comfortable with certain levels of technical debt in certain functionality, and should be free to make decisions as appropriate.

To give a specific example, ingest removed detailed auth plans from their phase 1 roadmap because it incurred considerable cost (due to complex ingest ownership requirements) without delivering proportionate value. Auth* was inserted as a DCP wide pilot requirement, which created scope creep for ingest without adding value to the ingest MVP (as ingest had already created a mitigation plan).

There's a delicate balance to strike between an idealised architecture RFC and a lowest common denominator RFC that doesn't create an imposition on any charter. From this document, I'm unclear on what the process is for determining what technical debt is in scope for the architecture group and what is in scope for individual charters / tech leads.

Comment (mweiden) [reply charters/Architecture/charter.md]: @tburdett I will clarify this as `technical debt that impacts delivery of DCP as an integrated system and must be addressed by PM product plans...`.

Here's my reasoning:
1. Tech debt will occur between components, not just within them.
2. It's totally reasonable for teams to have their own thresholds for tech debt so long as individual tech debt doesn't keep DCP from delivering on its goals as an integrated system. If that debt does slow the whole project down, then it is, by nature, a group problem.

File: charters/Architecture/charter.md
Comment (bkmartinjr) [inline charters/Architecture/charter.md]: As I understand the scope of this team, it comprises all (software) technology oversight, including:
* design oversight (aka architecture)
* operational policy oversight (see DevSecOps item below)
* implementation policy oversight, especially end-to-end (see Integration or Technical Best Practices items below)

If correct, the first sentence should capture the full scope.

Suggested revision:
> The Data Coordination Platform (DCP) Architecture Team is responsible for the technical oversight of the DCP, including identifying and addressing *cross-cutting* software design, implementation and operational concerns across DCP projects.   

I also suggest you organize the strategic and tactical responsibilities (below) by these domains.   As workload and complexity increase, some of them may naturally be delegated to sub-teams.

File: charters/Architecture/charter.md
Comment (bkmartinjr) [inline charters/Architecture/charter.md]: ... including design, implementation and operational policies

File: charters/Architecture/charter.md
Comment (bkmartinjr) [inline charters/Architecture/charter.md]: what is "technology radar"?   Is this a well-known term (a quick google search didn't help)...

File: charters/Architecture/charter.md
Comment (bkmartinjr) [inline charters/Architecture/charter.md]: Possibly missing strategic responsibility:  technology input on the PM roadmap process.   

Example: ensuring that technical dept is accounted in roadmapping/planning processes.

File: charters/Architecture/charter.md
Comment (bkmartinjr) [inline charters/Architecture/charter.md]: Need technical leads from non-component (ie, non-implementation) teams that could intersect with the "technology oversight" role, eg,
* devsecops team lead
* system integration / release / CI/CD team lead
* schema/metadata team team
etc.

File: charters/Architecture/charter.md
Comment (bkmartinjr) [inline charters/Architecture/charter.md]: Release processes & criteria?     (or is this included in "ways of work")?

**Review by bkmartinjr — CHANGES_REQUESTED** [review]

File: charters/Architecture/charter.md
Comment (mweiden) [reply charters/Architecture/charter.md]: https://www.thoughtworks.com/radar/faq companies publish these internally too. We did at SoundCloud.

I'm removing this, however. A Tech Radar is an format that the published docs can take on. There are probably others that would work just as well.

File: charters/Architecture/charter.md
Comment (bkmartinjr) [inline charters/Architecture/charter.md]: nit: what is an "RFC design"?   Perhaps "... on all RFCs"?

**Review by bkmartinjr — APPROVED** [review]

File: charters/Architecture/charter.md
Comment (brianraymor) [reply charters/Architecture/charter.md]: Might also clarify that the charter scope is for **technical* RFCs.

File: charters/Architecture/charter.md
Comment (tburdett) [inline charters/Architecture/charter.md]: I think the specifics here (technical debt, DCP as an integrated system) aren't needed. The arch team role will go beyond this quite narrow scope, and will publish RFCs for any major change. I'd suggest 
* Publish Requests for Comments (RFCs) for any significant change to the DCP architecture

**Review by tburdett — APPROVED** [review]
In general I approve this request, with one very minor comment

File: charters/Architecture/charter.md
Comment (tburdett) [inline charters/Architecture/charter.md:14]: You could add something here about the architecture team taking responsibility for having an overview and understanding of DCP-wide end to end requirements

File: charters/Architecture/charter.md
Comment (mweiden) [reply charters/Architecture/charter.md]: I think there is a misunderstanding here. This point is about making tech debt visible, not the about the format in which we communicate desired change (RFCs).

Also, I think saying Arch is the source of RFCs would change the core responsibilities in the role. Currently, Arch's role is framed as advisory and mentorship, acknowledging that we will (hopefully) receive many submissions from the community.

Comment (tburdett) [reply charters/Architecture/charter.md]: OK, so leading with "Publish Architecture Requests for Comments (RFCs)..." is probably what caused me to misinterpret this. I agree that we shouldn't be discouraging community contributions, I definitely was not intending to suggest the arch group was the only source of RFCs! So let's try that again... 😄 

If you are attempting to highlight the advisory role of the arch group, I'd say "documenting technical debt" is a very passive way of doing so. Who are you expecting to read this documentation? I prefer proposal of solutions rather than identification of problems.

How about something like:
 - Identify risks impacting delivery of the DCP roadmap and propose appropriate mitigation strategies by publishing requests for comments (RFCs) ...?

File: charters/Architecture/charter.md
Comment (briandoconnor) [inline charters/Architecture/charter.md:18]: I'm wondering if the [PM group](https://github.com/HumanCellAtlas/dcp-community/pull/23) should approve the User Stories section of these?  I'm worried that we have a formal way to connect the technical development of DCP with the user needs.

File: charters/Architecture/charter.md
Comment (briandoconnor) [inline charters/Architecture/charter.md:23]: I think between this group and the PM group it's important to keep an eye on what other large consortium are doing.  NIH, EBI, etc all have major cloud infrastructure projects and, as a scientist, being able to work across these environments and access data from HCA, GDC, GTEx, HIVE HubMap, etc would be helpful.  We see this happening with commercial and academic platforms starting to lower barriers for their users to pull data from a different large projects and we'd want the same thing to happen for HCA, e.g. researchers can use HCA data on the cloud in their favorite platform and combine with other interesting datasets.  There are technical coordination that need to happen to make this easier/possible.  Because of this, I think the tech group should have an item in the charter for "monitor and participate in technical coordination with other large projects as needed to facilitate data use".  Or something similar...

**Review by briandoconnor — COMMENTED** [review]
I left a few comments...

File: charters/Architecture/charter.md
Comment (mweiden) [reply charters/Architecture/charter.md:18]: Is it not enough that technical charters go through an open round of review with PM before going to Tech Arch for approval?

File: charters/Architecture/charter.md
Comment (mweiden) [reply charters/Architecture/charter.md:23]: Resolved with new bullet

**Pull request merged**
